### PR TITLE
Falcon7b optimize prefill and refactor

### PR DESCRIPTION
**Pull request by djordje-tt**
Code contains couple of features: 
- Split decode and prefill implementations of attn and mlp into subclasses
    - prefill and decode reuse weights between object for mlp and attn
    - 
- Optimize Falcon LM head for seq_len > 512
- Optimize prefill for seq_lens 128,1024 and 2048

### CODE REVIEW

**Review by arakhmati — APPROVED** [review]

**Comment by skhorasganiTT** [comment]
Can you please make sure that all of the test files under falcon7b/tests are passing with these changes? Currently in the PR there are no modifications to `test_falcon_model.py`, `test_falcon_prefill_decode.py`, `test_perf_falcon.py` and I believe they will need to be modified. 

Also, is the [models perf pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) passing with these changes?

File: models/demos/falcon7b/tt/falcon_model.py [hunk 1]
```diff
@@ -98,7 +95,49 @@ def __init__(
...
+                tt_memory_config=self.model_config["ATTN_MASK_MEMCFG"],
+                tt_dtype=self.model_config["ATTN_MASK_DTYPE"],
+            )
+            attn_masks_per_slice.append(tt_attention_slices)
+
+            attention_mask_starting_index_per_slice = (
+                attention_mask_starting_index_per_slice + number_of_attention_mask_elements_used_per_slice
+            ) % seq_len
+
+        return attn_masks_per_slice
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tt/falcon_model.py:138]: Can `_create_prefill_attn_mask_for_sharded_softmax` be commonized as much as possible with the near-duplicate version of the function in `model_utils.py` and imported from there?

File: models/demos/falcon7b/tt/falcon_model.py [hunk 2]
```diff
@@ -109,9 +148,18 @@ def model_preprocessing(self, llm_mode, input_ids, kv_cache_len, num_input_token
...
+                tt_input_ids.append(
+                    ttnn.as_tensor(
+                        input_ids[i : i + 1],
+                        dtype=self.model_config["INPUT_DTYPE"],
+                        layout=ttnn.ROW_MAJOR_LAYOUT,
+                        device=device,
+                        memory_config=self.model_config["INPUT_MEMCFG"],
+                    )
+                )
+
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tt/falcon_model.py:162]: The changes to this function seem to mistakenly be reverting some recent changes [539750a]((https://github.com/tenstorrent/tt-metal/commit/539750a91d6545a17df69f26dd1185d4611be2f6)) for async optimizations

File: models/demos/falcon7b/tt/falcon_mlp.py [hunk 3]
```diff
@@ -47,11 +66,236 @@ def __init__(
...
         for device_id in range(len(x)):
+            # pad inputs with padding tensor if not already padded
+            if (
+                x[device_id].shape[-1] < self.padding_value
+                and self.prefill_seq_len in [1024, 2048]
+                and self.model_config["OPTIMIZED_MODE"]
+            ):
+                x[device_id] = ttnn.concat(
+                    [x[device_id], self.model_config["MLP_DECODE_PADDING_TENSORS"][device_id]], dim=3
+                )
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tt/falcon_mlp.py:298]: Does this padding affect decode perf? Could we not instead unpad the weights and do the MLP as before for decode?

File: models/demos/falcon7b/tt/falcon_attention.py [hunk 4]
```diff
@@ -317,29 +747,16 @@ def forward(
                     -2,
                     -1,
                     output_mem_config=(
-                        self.model_config["K_TRANSPOSED_OUTPUT_MEMCFG"]
-                        if llm_mode == "prefill" or self.model_config["l1_sharded"] == False
-                        else tt_lib.tensor.MemoryConfig(
+                        tt_lib.tensor.MemoryConfig(
                             tt_lib.tensor.TensorMemoryLayout.HEIGHT_SHARDED, tt_lib.tensor.BufferType.L1
                         )
```
Comment (skhorasganiTT) [inline models/demos/falcon7b/tt/falcon_attention.py:752]: this should still be set to `self.model_config["K_TRANSPOSED_OUTPUT_MEMCFG"]` if `self.model_config["l1_sharded"] == False` right?

**Review by skhorasganiTT — CHANGES_REQUESTED** [review]

File: models/demos/falcon7b/tt/falcon_mlp.py (hunk 3 above)
Comment (djordje-tt) [reply models/demos/falcon7b/tt/falcon_mlp.py:298]: Concat and unpad add 11us of overhead so I believe it doesn't affect decode perf. On the other hand, it helps us optimize prefill MLP matmuls with convenient weight dims (for sharded params) and that's why we pad MLP weights. Here are unit test perf numbers:
[ops_perf_results_2024_04_25_12_46_16.csv](https://github.com/tenstorrent/tt-metal/files/15110400/ops_perf_results_2024_04_25_12_46_16.csv)

**Comment by djordje-tt** [comment]
Tests on t3000 and all perf tests are fine: 
perf run: https://github.com/tenstorrent/tt-metal/actions/runs/8850635801
post commit: https://github.com/tenstorrent/tt-metal/actions/runs/8850890544

**Review by skhorasganiTT — APPROVED** [review]
